client/mixnet/src/packet_dispatcher.rs

```diff
@@ -0,0 +1,207 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! [`AddressedPacket`] dispatching.
+
+use arrayvec::ArrayVec;
+use libp2p_identity::PeerId;
+use log::{debug, error, warn};
+use mixnet::{
+	core::{AddressedPacket, NetworkStatus, Packet, PeerId as CorePeerId},
+	network::{from_core_peer_id, to_core_peer_id, INVALID_CORE_PEER_ID},
+};
+use parking_lot::Mutex;
+use sc_network::{NetworkNotification, ProtocolName};
+use std::{collections::HashMap, future::Future, sync::Arc};
+
+const LOG_TARGET: &str = "mixnet";
+
+/// Packet queue for a peer.
+///
+/// Ideally we would use `Rc<RefCell<_>>`, but that would prevent the top-level future from being
+/// automatically marked `Send`. I believe it would be safe to manually mark it `Send`, but using
+/// `Arc<Mutex<_>>` here is not really a big deal.
+struct PeerQueue(Mutex<ArrayVec<Box<Packet>, 2>>);
+
+impl PeerQueue {
+	fn new() -> Self {
+		Self(Mutex::new(ArrayVec::new()))
+	}
+
+	/// Push `packet` onto the queue. Returns `true` if the queue was previously empty. Fails if
+	/// the queue is full.
+	fn push(&self, packet: Box<Packet>) -> Result<bool, ()> {
+		let mut queue = self.0.lock();
+		if queue.is_full() {
+			Err(())
+		} else {
+			let was_empty = queue.is_empty();
+			queue.push(packet);
+			Ok(was_empty)
+		}
+	}
+
+	/// Drop all packets from the queue.
+	fn clear(&self) {
+		let mut queue = self.0.lock();
+		queue.clear();
+	}
+
+	/// Pop the packet at the head of the queue and return it, or, if the queue is empty, return
+	/// `None`. Also returns `true` if there are more packets in the queue.
+	fn pop(&self) -> (Option<Box<Packet>>, bool) {
+		let mut queue = self.0.lock();
+		let packet = queue.pop();
+		(packet, !queue.is_empty())
+	}
+}
+
+/// A peer which has packets ready to send but is not currently being serviced.
+pub struct ReadyPeer {
+	id: PeerId,
+	/// The peer's packet queue. Not empty.
+	queue: Arc<PeerQueue>,
```

Comment:
A `ReadyPeer` doesn't live very long, but the `PeerQueue` reference does; it is stored in the future returned by `ReadyPeer::send_packet`. This future is stuffed in a `FuturesUnordered` and may still be alive when `PacketDispatcher::dispatch` is called again. `dispatch` may push more packets into the queue. A simple borrow of the queue is not sufficient.

As the comment above `PeerQueue` says, `Arc<Mutex<_>>` isn't really needed, as everything is effectively single-threaded; `Rc<RefCell<_>>` would do. I don't think it would be possible to have the `run` future be `Send` using `Rc<RefCell<_>>` though, without using `unsafe`. `Arc<Mutex<_>>` isn't that expensive when there is no contention, so I wasn't particularly worried about this.

Reply:
Ahh, I had `Ok(more_packets) => more_packets.then(|| self)` overseen. Yeah, then my comment is useless^^